### DISCUSSION TITLE
Thermal Offset

### DISCUSSION THREAD

**Discussion by cyring** [cyring/CoreFreq#367] [Show and tell]
Starting with commit 9eaf4c55fc31d9d53e179f23ae172e4438bec289 of `develop` branch, one can select a temperature offset.

This is architectural, you have to refer to your processor datasheet.

The _CoreFreq_ driver can accept or reject a request if the offset is not within the specified bounds or if there is no implementation (so far Intel only).

# Preview: Tiger Lake/U
* Adding 5 degrees to the offset
* You will observe the resulted Core's temperature
![2022-10-30-094743_644x508_scrot](https://user-images.githubusercontent.com/11563789/198870311-f424af51-b013-4e51-b942-b560fe27b658.png)
* Offset is now `15 C`
* Decreasing 9 degrees
![2022-10-30-094820_644x508_scrot](https://user-images.githubusercontent.com/11563789/198870313-b194f083-cf41-4153-89a5-405b9030cb20.png)
* Offset is now `6 C`
![2022-10-30-094833_644x508_scrot](https://user-images.githubusercontent.com/11563789/198870316-25d18718-c3bd-45e8-985c-8268a08ef367.png)
* Attempting an increase of `+20 C`
![2022-10-30-100136_644x508_scrot](https://user-images.githubusercontent.com/11563789/198870635-aff6daf4-f574-441c-b55b-412b38462571.png)
* Request rejected: the computed offset must remain within `[0, +15]`
![2022-10-30-100236_644x508_scrot](https://user-images.githubusercontent.com/11563789/198870679-2dcc5b1d-4a29-44d4-a72a-5661dfe0a230.png)

**Comment by BugReporterZ** [comment]
On my i7-12700K any offset equal or lower than 0 is causing an "Invalid argument" error, while higher values do not seem to work at all.

Note that otherwise on my motherboard it is possible to set in BIOS any TjMax value in the 65-115 °C range; currently I have it at 85 °C. I tried configuring it from CoreFreq also after setting it from BIOS to the default value of 100 °C, but with the same results.

**Reply by cyring** [reply]
Thanks for trying. I would like to debug from the Thermal Register. Can you read as below:

```sh
## MSR_IA32_TEMPERATURE_TARGET
rdmsr -ax 0x1A2
```
Offset value should be within bits `[28-24]`
If not, then no Alder Lake support

**Reply by BugReporterZ** [reply]
I get:

```
# rdmsr -ax 0x1A2
CPU 0: 80551400
CPU 1: 80551400
CPU 2: 80551400
CPU 3: 80551400
CPU 4: 80551400
CPU 5: 80551400
CPU 6: 80551400
CPU 7: 80551400
CPU 8: 80551400
CPU 9: 80551400
CPU 10: 80551400
CPU 11: 80551400
CPU 12: 80551400
CPU 13: 80551400
CPU 14: 80551400
CPU 15: 80551400
CPU 16: 80551400
CPU 17: 80551400
CPU 18: 80551400
CPU 19: 80551400
```

Bits [23-16] above should be "55" (hex) -> 85 (dec) which is my current TjMax.
Bits [27-24] above should be "0". Indeed my TjMax offset is shown as 0.

![image](https://user-images.githubusercontent.com/26941368/198882390-3375f754-de29-4fee-9ebd-88e54c99feeb.png)

**Reply by cyring** [reply]
Starting from zero, offset can be selected between `+1` and `+15`. The more you add, the lowest temperature you should get. 
This is what I have observed with Tiger Lake; you have to tell me about the effects on Alder Lake. 

Any minus value will bring back a computed offset to zero, meaning no temperature offset is programmed. Can't be negative. 


I can't tell which Registers BIOS is altering. But maybe more than 4 bits are used on Alder Lake, like some Atom whom field is 6 bits wide. See [definitions](https://github.com/cyring/CoreFreq/blob/9eaf4c55fc31d9d53e179f23ae172e4438bec289/intel_reg.h#L1444) I've gather so far. 

I would suggest you set the highest value in BIOS then `rdmsr -ax 0x1A2` to reverse the encoding.

**Reply by BugReporterZ** [reply]
I set the maximum value in BIOS (115 °C) and this is the result. The relevant bits changed to 73 (hex) -> 115 °C (dec):

```
# sudo rdmsr -ax 0x1A2
CPU 0: 80731400
CPU 1: 80731400
CPU 2: 80731400
CPU 3: 80731400
CPU 4: 80731400
CPU 5: 80731400
CPU 6: 80731400
CPU 7: 80731400
CPU 8: 80731400
CPU 9: 80731400
CPU 10: 80731400
CPU 11: 80731400
CPU 12: 80731400
CPU 13: 80731400
CPU 14: 80731400
CPU 15: 80731400
CPU 16: 80731400
CPU 17: 80731400
CPU 18: 80731400
CPU 19: 80731400
```

I then tried to set the offset via Corefreq to +15, but nothing changed, and the MSR reported the same values as well.
While +15 is silently accepted (doing nothing), setting +16 gives instead "Invalid Argument".

**Reply by cyring** [reply]
The upper bits `0x80XXXXXX` are thus meaningless and none `Target Offset` nor `TCC Activation Offset` is programmable on your ADL

![2022-10-30-165948_756x261_scrot](https://user-images.githubusercontent.com/11563789/198888689-9951a8c9-980c-41d5-83bb-5e2716793815.png)
![2022-10-30-170055_757x230_scrot](https://user-images.githubusercontent.com/11563789/198888693-8a1835aa-57b5-4fe5-8391-b40158d9d732.png)

But can you program the TjMax ? 
_Although it is specified as read-only in all previous architectures_

Something like:
```sh
## attempt to set TjMax to 110 C
wrmsr -a 0x1A2 0x806E1400
rdmsr -ax 0x1A2
```

**Reply by BugReporterZ** [reply]
It looks like it isn't possible to directly program TjMax in the OS.

```
# sudo wrmsr -a 0x1A2 0x806E1400
# sudo rdmsr -ax 0x1A2
CPU 0: 80551400
CPU 1: 80551400
CPU 2: 80551400
CPU 3: 80551400
CPU 4: 80551400
CPU 5: 80551400
CPU 6: 80551400
CPU 7: 80551400
CPU 8: 80551400
CPU 9: 80551400
CPU 10: 80551400
CPU 11: 80551400
CPU 12: 80551400
CPU 13: 80551400
CPU 14: 80551400
CPU 15: 80551400
CPU 16: 80551400
CPU 17: 80551400
CPU 18: 80551400
CPU 19: 80551400
```

**Reply by cyring** [reply]
Thanks I was expecting those results.

To sum up: my new feature "Thermal Offset" is working for some processors, at least those specified in the Intel SDM; and my Tiger Lake for Laptop for unknown reasons 🤔

**Reply by BugReporterZ** [reply]
It might be another of those laptop-specific features that aren't enabled on desktop parts.

**Comment by cyring** [comment]
# Preview: IvyBridge

* Offset has a nominal zero value
* Let's add `+10 C`

![2022-10-31-080953_644x508_scrot](https://user-images.githubusercontent.com/11563789/198951838-ba4cf464-cb57-43be-9ba9-d3f350a9cb66.png)

* Let's add `+5 C` more

![2022-10-31-081023_644x508_scrot](https://user-images.githubusercontent.com/11563789/198951906-e93f62ed-e9b5-4a05-bcf2-028e22806235.png)

* With a `15 C` offset, Core temperature computes to `12 C`

![2022-10-31-081542_644x508_scrot](https://user-images.githubusercontent.com/11563789/198952248-9ede7663-1f1c-4cbd-a74a-fa0f0af689c6.png)

* Let's roll back to a null Thermal Offset
* Apply a minus `15 C` to the offset

![2022-10-31-081944_644x508_scrot](https://user-images.githubusercontent.com/11563789/198952890-d73f92f5-f0da-4e20-84c6-fc03d9086702.png)

* Temperature = TjMax - DTS - offset

![2022-10-31-082041_644x508_scrot](https://user-images.githubusercontent.com/11563789/198953040-1683ec3f-d0a7-4395-a07d-29241eedea25.png)

**Comment by cyring** [comment]
# Preview: Westmere/Gulftown [Nehalem]
* Including commit d944c04582eee6ffd05a5d77dc56f004df03c493 , attempt to add `5 C` to the offset

![2022-10-31-092536_644x508_scrot](https://user-images.githubusercontent.com/11563789/198964447-b0e54d41-d2a5-4906-905d-4bcd6318aa8a.png)

* Processor does not support Thermal Offset from `MSR_IA32_TEMPERATURE_TARGET`

![2022-10-31-092559_644x508_scrot](https://user-images.githubusercontent.com/11563789/198964674-ed1c3f64-42ac-4da7-b513-e9f225222401.png)
